[ConstraintSystem] Add the notion of disjunction partitioning.

rudkx opened the pull request:
Update DisjunctionChoiceProducer so that it creates a sorted ordering
of the choices in the disjunction as well as a partitioning of those
choices.

If we hit the end of a partition and have found a solution, we'll stop
attempting choices from this disjunction.

For now we'll default to keeping the same order that the disjunctions
initially have, and will put them all into the same partition. A
future commit will implement logic to implement more interesting
partitions of the choices.

Commit:
[ConstraintSystem] Add the notion of disjunction partitioning.

Update DisjunctionChoiceProducer so that it creates a sorted ordering
of the choices in the disjunction as well as a partitioning of those
choices.

If we hit the end of a partition and have found a solution, we'll stop
attempting choices from this disjunction.

For now we'll default to keeping the same order that the disjunctions
initially have, and will put them all into the same partition. A
future commit will implement logic to implement more interesting
partitions of the choices.

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci Please test source compatibility

rudkx commented:
@xedin What do you think of this approach to trying groups of disjunctions and stopping as soon as we find a group that successfully solves the system?

xedin commented on lib/Sema/ConstraintSystem.h:
```diff
@@ -3516,22 +3543,51 @@ class DisjunctionChoiceProducer : public BindingProducer<DisjunctionChoice> {
+
+    bool endOfPartition = false;
+
+    if (PartitionEnd[PartitionIndex] == Index) {
```
Maybe instead of `isEndOfDisjunctionPartition` it would be better to rename `needsToComputeNext()` into `endOfPartitition()` and use it for both producers? That way `BindingStep` could just implement non-virtual `shouldStopAfter` as `AnySolved && Producer.endOfPartition()` and this `if` block could be moved there, WDYT?

rudkx replied:
Sure. I'll push a change for that.

xedin commented:
@rudkx I think that approach makes sense, if there is a criteria which would be used to partition them!

Commit:
[ConstraintSystem] Make shouldStopAfter non-virtual.

Rename `needsToComputeNext()` to `isEndOfPartition()` and have
`shouldStopAfter()` test `AnySolved() && isEndOfPartition()` to
determine if we can stop iterating over choices for the current step.

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci Please test source compatibility

xedin reviewed (approved):
LGTM!

rudkx commented:
@swift-ci Please smoke test OS X platform

rudkx commented:
@swift-ci Please test source compatibility

rudkx commented:
Stepping through this I'm realizing that I don't want to use `shouldStopAfter()`, which only appears to get called on failures, but rather `shouldStopAt()`, which is called before the attempt, so I'll take a look at that.

xedin commented:
Just FYI, the difference between `shouldStopAfter` and `shouldStopAt` is that the latter would cause at least one binding from the next partition to be produced, where former is going to stop exactly at the end of the partition.

Commit:
Revert "[ConstraintSystem] Make shouldStopAfter non-virtual."

This reverts commit bccb44954422dec079a422271fbfc59e66199193.

Commit:
[ConstraintSystem] Use shouldStopAt to limit disjunction choices.

Rather than using shouldStopAfter, use shouldStopAt, and check if
we're at the beginning of a disjunction (with an already successful
solution).

rudkx commented:
@swift-ci Please smoke test

rudkx commented:
@swift-ci Please test source compatibility
